src/util/key.rs

```diff
@@ -11,86 +11,94 @@
+    }
+
+    /// deserializes it from a slice
+    pub fn from_slice<C>(secp: &Secp256k1<C>, data: &[u8]) -> Result<PublicKey, encode::Error> {
+        let key: secp256k1::PublicKey = secp256k1::PublicKey::from_slice(&secp, data)
+            .map_err(|_| base58::Error::Other("Public key out of range".to_owned()))?;
+
+        let compressed: bool = match data.len() {
+            33 => true,
+            65 => false,
+            _ =>  { return Err(base58::Error::InvalidLength(data.len()).into()); },
+        };
+
+        Ok(PublicKey {
             compressed: compressed,
-            network: network,
             key: key,
+        })
+    }
+
+    /// Computes the public key as supposed to be used with this secret
+    pub fn from_private_key<C: secp256k1::Signing>(secp: &Secp256k1<C>, sk: &PrivateKey) -> PublicKey {
+        PublicKey {
+            compressed: sk.compressed,
+            key: secp256k1::PublicKey::from_secret_key(secp, &sk.key)
         }
     }
 
+}
+
+#[derive(Clone, PartialEq, Eq)]
+/// A Bitcoin ECDSA private key
+pub struct PrivateKey {
+    /// Whether this private key represents a compressed address
+    pub compressed: bool,
+    /// The network on which this key should be used
+    pub network: Network,
+    /// The actual ECDSA key
+    pub key: secp256k1::SecretKey,
+}
+
+impl PrivateKey {
     /// Computes the public key as supposed to be used with this secret
     pub fn public_key<C: secp256k1::Signing>(&self, secp: &Secp256k1<C>) -> PublicKey {
-        PublicKey::from_secret_key(secp, &self.key)
+        PublicKey {
+            compressed: self.compressed,
+            key: secp256k1::PublicKey::from_secret_key(secp, &self.key)
+        }
     }
 
     /// Converts a private key to a segwit address
-    #[inline]
     pub fn to_address<C: secp256k1::Signing>(&self, secp: &Secp256k1<C>) -> Address {
-        Address::p2wpkh(&self.public_key(secp), self.network)
+        Address::p2wpkh(self.public_key(secp), self.network)
     }
 
     /// Converts a private key to a legacy (non-segwit) address
-    #[inline]
     pub fn to_legacy_address<C: secp256k1::Signing>(&self, secp: &Secp256k1<C>) -> Address {
-        if self.compressed {
-            Address::p2pkh(&self.public_key(secp), self.network)
-        }
-        else {
-            Address::p2upkh(&self.public_key(secp), self.network)
-        }
-    }
-
-    /// Accessor for the underlying secp key
-    #[inline]
-    pub fn secret_key(&self) -> &SecretKey {
-        &self.key
-    }
-
-    /// Accessor for the underlying secp key that consumes the privkey
-    #[inline]
-    pub fn into_secret_key(self) -> SecretKey {
-        self.key
-    }
-
-    /// Accessor for the network type
-    #[inline]
-    pub fn network(&self) -> Network {
-        self.network
-    }
-
-    /// Accessor for the compressed flag
-    #[inline]
-    pub fn is_compressed(&self) -> bool {
-        self.compressed
+        Address::p2pkh(self.public_key(secp), self.network)
```

Comment:
Given the amount of ways you can go from a key to an address, I think it might be better to just have the explicit `Address` constructors and no getters here at all.

Reply:
What do you mean by this? Do you mean that we should eliminate `PrivateKey::to_{legacy_,}address`?

Reply:
Yeah. I mean there is p2pk, p2pkh, p2wpkh. Which one is "address" and which one is "legacy address"? Every time there is a new way to make an address from a key, this will get more confusing. The `Address` constructors are explicit and very clear.

If you really care about the getters on the key, I'd make them more explicit. Like `to_p2pkh_address`, ... But I think the constructors are sufficient.

Reply:
@stevenroose Agreed.

Reply:
@stevenroose Actually... I've thought about it a little... It's a little un-ergonomic that we would have the use the constructor like so: `Address::p2pk(sk.public_key(&secp), sk.network)`.

Perhaps we _should_ do `PrivateKey::to_p2pkh_address(&secp)` etc., just for ergonomics sake.

Reply:
Hmm, didn't notice that, but it is strange to me that a secret key has a network attached. And yeah the secp context is bothersome, but will soon go away for (de)serialization I heard (@apoelstra?). So well, `Address::p2pk(sk.public_key(), sk.network)` doesn't seem like that bad to me.

But yeah I'd be ok with the address getters as long as they are explicit about what type of address they generate.

Reply:
@apoelstra But just for WIF's sake?

Reply:
@stevenroose correct

Reply:
I feel like that's kind of a pity. I mean WIF is just an "import format", right? I feel like it would be perfectly acceptable to have wif just as a `util/wif.rs` thing for compatibility sake if it's actually being annoying wrt other things.

Reply:
A bit like how an Address is different from a PublicKey, a Wif could be different from a PrivateKey.

Reply:
We could do that. It would mean that you couldn't serialize a `PrivateKey` without converting to a `WIf` first, and you couldn't do that without knowing your network (which you'd need to obtain from somewhere)...and you'd then be unable to obtain an address from a `PrivateKey` without doing the same thing ... so the `PrivateKey` type becomes very unergonomic.

Reply:
Yeah it becomes basically useless because of `secp256k1::key::SecretKey`. But that's not necessarily a problem. `Address::p2pkh(secp_priv.public_key(), network)` doesn't seem wrong though (considering context-less priv-pub conversion). 
And you really only need WIF strings when you actually want WIF strings, no? `bitcoin::utils::wif::to_wif(my_pk, network)` also not.

Question: will the secp256k1 `SecretKey` and `PublicKey` look different for Schnorr?